crates/runtime/src/component/dataset/acceleration.rs

```diff
@@ -0,0 +1,320 @@
+pub enum OnConflictBehavior {
+    #[default]
+    Drop,
+    Upsert,
+}
+
+impl From<spicepod_acceleration::OnConflictBehavior> for OnConflictBehavior {
+    fn from(index_type: spicepod_acceleration::OnConflictBehavior) -> Self {
+        match index_type {
+            spicepod_acceleration::OnConflictBehavior::Drop => OnConflictBehavior::Drop,
+            spicepod_acceleration::OnConflictBehavior::Upsert => OnConflictBehavior::Upsert,
+        }
+    }
+}
+
+impl From<&str> for OnConflictBehavior {
+    fn from(index_type: &str) -> Self {
+        match index_type.to_lowercase().as_str() {
+            "upsert" => OnConflictBehavior::Upsert,
+            _ => OnConflictBehavior::Drop,
+        }
+    }
+}
+
+impl Display for OnConflictBehavior {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            OnConflictBehavior::Drop => write!(f, "drop"),
+            OnConflictBehavior::Upsert => write!(f, "upsert"),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Acceleration {
+    pub enabled: bool,
+
+    pub mode: Mode,
+
+    pub engine: Engine,
+
+    pub refresh_mode: RefreshMode,
+
+    pub refresh_check_interval: Option<String>,
+
+    pub refresh_sql: Option<String>,
+
+    pub refresh_data_window: Option<String>,
+
+    pub params: HashMap<String, String>,
+
+    pub engine_secret: Option<String>,
+
+    pub retention_period: Option<String>,
+
+    pub retention_check_interval: Option<String>,
+
+    pub retention_check_enabled: bool,
+
+    pub on_zero_results: ZeroResultsAction,
+
+    pub indexes: HashMap<ColumnReference, IndexType>,
+
+    pub primary_key: Option<ColumnReference>,
+
+    pub on_conflict: HashMap<ColumnReference, OnConflictBehavior>,
+}
+
+impl TryFrom<spicepod_acceleration::Acceleration> for Acceleration {
+    type Error = crate::Error;
+
+    fn try_from(
+        acceleration: spicepod_acceleration::Acceleration,
+    ) -> std::result::Result<Self, Self::Error> {
+        let try_parse_column_reference = |column: &str| {
+            ColumnReference::try_from(column).map_err(|e| crate::Error::InvalidSpicepodDataset {
+                source: super::Error::UnableToParseColumnReference {
+                    column_ref: column.to_string(),
+                    source: e,
+                },
+            })
+        };
+
+        let primary_key = match acceleration.primary_key {
+            Some(pk) => Some(try_parse_column_reference(pk.as_str())?),
+            None => None,
+        };
+
+        let mut indexes = HashMap::new();
+        for (k, v) in acceleration.indexes {
+            indexes.insert(try_parse_column_reference(k.as_str())?, IndexType::from(v));
+        }
+
+        let mut on_conflict = HashMap::new();
+        for (k, v) in acceleration.on_conflict {
+            on_conflict.insert(
+                try_parse_column_reference(k.as_str())?,
+                OnConflictBehavior::from(v),
+            );
+        }
```

Comment:
We parse all of the column references up front when we're translating from the Spicepod representation to the runtime representation. This front-loads any errors that might come up from invalid configuration, and also makes it much easier to work with later on.